python-lib/ontology_tagging/formatter_by_match.py

```diff
@@ -0,0 +1,165 @@
+        logging.info(
+            f"Tagging {len(input_df)} documents : Done in {perf_counter() - start:.2f} seconds."
+        )
+        return self._set_columns_order(self._duplicate_df, self.output_df, text_column)
+
+    def _write_row(self, row: pd.Series, language_column: AnyStr = None) -> None:
+        """
+        Called by write_df on each row
+        Update the output dataframes which will be concatenated after that:
+        - self.output_df is filled with columns with the tag columns
+        - self._duplicate_df is filled with the original rows of the Document Dataset, with copies
+        There are as many copies of a document as there are keywords in this document
+        Args:
+            row: pandas.Series from text_df
+            language_column: if not None, matcher will apply with the given language of the row
+        """
+        self.contains_match = False
+        language = self._get_document_language(row, language_column)
+        matches = []
+        document_to_match = self._get_document_to_match(row, language)
+        empty_row = {column: np.nan for column in self.tag_columns}
+        if not self.category_column:
+            matches = [
+                (
+                    self._matcher_dict[language](sentence, as_spans=True),
+                    row[self.text_column_tokenized][idx],
+                )
+                for idx, sentence in enumerate(document_to_match)
+            ]
+            self._get_tags_in_row(matches, row, language)
+        else:
+            self._get_tags_in_row_category(document_to_match, row, language)
+        if not self.contains_match:
+            self.output_df = self.output_df.append(empty_row, ignore_index=True)
+            self._duplicate_df = self._duplicate_df.append(
+                pd.DataFrame([row]), ignore_index=True
+            )
+
+    def _get_tags_in_row(self, matches: List, row: pd.Series, language: AnyStr) -> None:
+        """
+        Called by _write_row
+        Create the list of new rows with infos about the tags and gives it to _update_df
+        """
+        values = []
+        for match, sentence in matches:
+            values = [
+                self._list_to_dict(
+                    [
+                        self._keyword_to_tag[language][
+                            get_span_text(span=keyword, lemmatize=self.lemmatization)
+                        ],
+                        keyword.text,
+                        sentence,
+                    ]
+                )
+                for keyword in match
+            ]  # new rows to append to the output dataframe
+            self._update_df(match, values, row)
+
+    def _get_tags_in_row_category(
+        self, document_to_match: List, row: pd.Series, language: AnyStr
+    ) -> None:
+        """
+        Called by _write_row_category
+        Create the list of new rows with infos about the tags and gives it to _update_df function
+        """
+        tag_rows = []
```

Comment:
ditto, no need to initialize

Reply:
ditto, https://github.com/dataiku/dss-plugin-nlp-analysis/pull/20#discussion_r639838168